### PR TITLE
Bluetooth: Mesh: Add key-value concept to store model data

### PR DESCRIPTION
**Pull request by LingaoM**
Previous mode store function only can store single data,
change this to store as KV model, let's app-layer to manager
model data, other than by stack when node reset.

Signed-off-by: Lingao Meng <mengabc1086@gmail.com>

### CODE REVIEW

File: include/bluetooth/mesh/access.h
Comment (trond-snekvik) [inline include/bluetooth/mesh/access.h:669]: This key should be a string as well, to stay consistent with the rest of the settings API, in my opinion. The max length of a single settings entry name is 7 characters, so if this function only accepts a single level of data, it shouldn't introduce any complexity. 

This also opens up for passing `key=NULL` to get the same single-entry functionality as the current API.

File: subsys/bluetooth/mesh/cfg_srv.c
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/cfg_srv.c:3323]: Add a note in the reset callback API?
For instance

`@note If the model stores any persistent data, this needs to be erased manually.`

File: subsys/bluetooth/mesh/settings.c
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/settings.c:2448]: This buffer needs to be extended to fit the additional path length.

File: subsys/bluetooth/mesh/settings.c
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/settings.c:2457]: The `MOD_DATA_PRESENT` flag doesn't make sense anymore. It's currently only used to keep track of whether data should be deleted, but this isn't universal anymore, since you might want to delete only one of the model data entries.

I think it can be removed, and lines 2457-2465 can be reduced to `err = settings_save_one(path, data, data_len);`.

**Review by trond-snekvik — CHANGES_REQUESTED** [review]
Overall, this looks like a good change. See comments.

File: subsys/bluetooth/mesh/settings.c
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/settings.c:2454]: Can go back to old code now to simplify this:

```C
encode_mod_path(mod, vnd, "data", path, sizeof(path));
if (name) {
    strcat(path, "/");
    strncat(path, name, 8);
}
```

**Review by trond-snekvik — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/mesh/settings.c
Comment (LingaoM) [reply subsys/bluetooth/mesh/settings.c:2454]: You're too smart

**Review by trond-snekvik — APPROVED** [review]
LGTM. @jhedberg this is an API change, does it need to be noted anywhere?

**Comment by jhedberg** [comment]
> LGTM. @jhedberg this is an API change, does it need to be noted anywhere?

Yes, I suspect it should be mentioned in the release notes for 2.4. @carlescufi should be able to give more guidance if necessary.

**Review by jhedberg — APPROVED** [review]

**Comment by carlescufi** [comment]
@LingaoM thanks! unfortunately this is a new feature and so it will need to wait until the feature merge window opens next week.

**Comment by carlescufi** [comment]
Regarding the API change, if this is a stable API then you need to follow the guidelines here:
https://docs.zephyrproject.org/latest/development_process/api_lifecycle.html#introducing-incompatible-changes

**Comment by LingaoM** [comment]
#26037

**Pull request merged**
